walk/walk.go

```diff
@@ -109,213 +110,379 @@ type WalkFunc func(dir, rel string, c *config.Config, update bool, f *rule.File,
-func (u *UpdateFilter) shouldVisit(rel string, updateParent bool) bool {
-	switch u.mode {
-	case VisitAllUpdateSubdirsMode, VisitAllUpdateDirsMode:
-		return true
-	case UpdateSubdirsMode:
-		_, ok := u.updateRels[rel]
-		return ok || updateParent
-	default: // UpdateDirsMode
-		_, ok := u.updateRels[rel]
-		return ok
+// visit is the main recursive function of walker. It visits one directory,
+// possibly recurses into subdirectories, and possible calls the callback.
+//
+// updateParent should indicate whether the the current mode tells Gazelle
+// to call the callback in the parent directory with update = true (see
+// shouldUpdate). The callback may not actually be called if the build file
+// contains syntax errors or a gazelle:ignore directive.
+func (w *walker) visit(c *config.Config, rel string, updateParent bool) {
+	w.visits[rel] = visitInfo{c: c} // to be updated when we're further along.
+
+	// Absolute path to the directory being visited
+	dir := filepath.Join(c.RepoRoot, rel)
+
+	// Load the build file.
+	info, err := w.cache.get(rel, w.loadDirInfo)
+	if err != nil {
+		w.errs = append(w.errs, err)
+	}
+	hasBuildFileError := err != nil
+
+	configure(w.cexts, w.knownDirectives, c, rel, info.file, info.config)
+	wc := info.config
+	regularFiles := info.regularFiles
+	subdirs := info.subdirs
+
+	if wc.isExcludedDir(rel) {
+		return
+	}
+
+	containedByParent := info.file == nil && wc.updateOnly
+	w.visits[rel] = visitInfo{
+		c:                 c,
+		containedByParent: containedByParent,
+		regularFiles:      regularFiles,
+		subdirs:           subdirs,
+	}
+
+	// Visit subdirectories, as needed.
+	shouldUpdate := w.shouldUpdate(rel, updateParent)
+	for _, subdir := range subdirs {
+		subdirRel := path.Join(rel, subdir)
+		if w.shouldVisit(subdirRel, info.config, shouldUpdate) {
+			w.visit(c.Clone(), subdirRel, shouldUpdate)
+			if c.Strict && len(w.errs) > 0 {
+				return
+			}
+		}
+	}
+
+	// Recursively collect regular files from subdirectories that won't contain
+	// build files. Files are added in depth-first pre-order.
+	if !containedByParent {
+		var collect func(string, string)
+		collect = func(rel, prefix string) {
+			vi := w.visits[rel]
+			if !vi.containedByParent {
+				return
+			}
+			for _, f := range vi.regularFiles {
+				regularFiles = append(regularFiles, path.Join(prefix, f))
+			}
+			for _, subdir := range vi.subdirs {
+				collect(path.Join(rel, subdir), path.Join(prefix, subdir))
+			}
+		}
+		for _, subdir := range subdirs {
+			collect(path.Join(rel, subdir), subdir)
+		}
+	}
+
+	// Call the callback to update this directory.
+	update := !wc.ignore && shouldUpdate && !hasBuildFileError
+	if w.shouldCall(rel, updateParent) {
+		genFiles := findGenFiles(wc, info.file)
+		result := w.wf(Walk2FuncArgs{
+			Dir:          dir,
+			Rel:          rel,
+			Config:       c,
+			Update:       update,
+			File:         info.file,
+			Subdirs:      subdirs,
```

Comment:
The `subdirs`s was not modified based on `containedByParent` like `regularFiles` was

Reply:
Fixed, but please double check